dice-where-downloader-lib/src/main/java/technology/dice/dicewhere/downloader/stream/StreamWithMD5Decorator.java

```diff
@@ -1,41 +1,82 @@
 package technology.dice.dicewhere.downloader.stream;
 
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
 import java.io.IOException;
 import java.io.InputStream;
 import java.security.DigestInputStream;
 import java.security.MessageDigest;
 import java.security.NoSuchAlgorithmException;
+import java.util.Optional;
 import javax.xml.bind.annotation.adapters.HexBinaryAdapter;
 import technology.dice.dicewhere.downloader.md5.MD5Checksum;
 
 public class StreamWithMD5Decorator extends InputStream {
 
+  private static final int BUFFER_SIZE = 8192;
+  private static final HexBinaryAdapter HEX_BINARY_ADAPTER = new HexBinaryAdapter();
+
   private final MessageDigest md5;
-  DigestInputStream inputStream;
+  private final DigestInputStream originalStream;
+
+  private Optional<InputStream> duplicatedStream = Optional.empty();
+  private Optional<MD5Checksum> md5Checksum = Optional.empty();
```

Comment:
so this class is stateful but without any documentation to that effect. Would be much cleaner if the only state was within the inputstream itself.